src/dc_chat.c

```diff
@@ -2517,21 +2475,137 @@ uint32_t dc_send_msg(dc_context_t* context, uint32_t chat_id, dc_msg_t* msg)
-		msg->id = send_msg_raw(context, chat, msg, dc_create_smeared_timestamp(context));
-		if (msg->id==0) {
-			goto cleanup; /* error already logged */
-		}
+		msg->id = prepare_msg_raw(context, chat, msg, dc_create_smeared_timestamp(context), state);
+		msg->chat_id = chat_id;
+		/* potential error already logged */
 	}
 
-	context->cb(context, DC_EVENT_MSGS_CHANGED, chat_id, msg->id);
-
 cleanup:
 	dc_chat_unref(chat);
 	free(pathNfilename);
 	return msg->id;
 }
 
 
+/**
+ * Prepare a message for sending.
+ *
+ * Call this function if the file to be sent is still in creation.
+ * Once you're done with creating the file, call dc_send_msg() with chat_id of 0
+ * and the message will really be sent.
+ *
+ * This is useful as the user can already send the next messages while
+ * e.g. the recoding of a video is not yet finished. Or the user can even forward
+ * the message with the file being still in creation to other groups.
+ *
+ * Files being sent with the increation-method must be placed in the
+ * blob directory, see dc_get_blobdir().
+ * If the increation-method is not used - which is probably the normal case -
+ * dc_send_msg() copies the file to the blob directory if it is not yet there.
+ *
+ * Example:
+ * ~~~
+ * dc_msg_t* msg = dc_msg_new(context, DC_MSG_VIDEO);
+ * dc_msg_set_file(msg, "/file/to/send.mp4", NULL);
+ * dc_prepare_msg(context, chat_id, msg);
+ * // ... after /file/to/send.mp4 is ready:
+ * dc_send_msg(context, 0, msg);
+ * ~~~
+ *
+ * @memberof dc_context_t
+ * @param context The context object as returned from dc_context_new().
+ * @param chat_id Chat ID to send the message to.
+ * @param msg Message object to send to the chat defined by the chat ID.
+ *     On succcess, msg_id of the object is set up,
+ *     The function does not take ownership of the object,
+ *     so you have to free it using dc_msg_unref() as usual.
+ * @return The ID of the message that is being prepared.
+ */
+uint32_t dc_prepare_msg(dc_context_t* context, uint32_t chat_id, dc_msg_t* msg)
+{
+	if (context==NULL || context->magic!=DC_CONTEXT_MAGIC || msg==NULL || chat_id<=DC_CHAT_ID_LAST_SPECIAL) {
+		return 0;
+	}
+
+	return prepare_msg_common(context, chat_id, msg, DC_STATE_OUT_PREPARING);
+}
+
+/**
+ * Send a message defined by a dc_msg_t object to a chat.
+ *
+ * Sends the event #DC_EVENT_MSGS_CHANGED on succcess.
+ * However, this does not imply, the message really reached the recipient -
+ * sending may be delayed eg. due to network problems. However, from your
+ * view, you're done with the message. Sooner or later it will find its way.
+ *
+ * Example:
+ * ~~~
+ * dc_msg_t* msg = dc_msg_new(context, DC_MSG_IMAGE);
+ * dc_msg_set_file(msg, "/file/to/send.jpg", NULL);
+ * dc_send_msg(context, chat_id, msg);
+ * ~~~
+ *
+ * @memberof dc_context_t
+ * @param context The context object as returned from dc_context_new().
+ * @param chat_id Chat ID to send the message to.
+ *     If dc_prepare_msg() was called before, this parameter must be 0.
```

Comment:
why not allow passing the chat_id even in this case?
imo this would provide a clearer api with less confusion.

i think we can check the prepare-state using `msg->state==DC_STATE_OUT_PREPARING` (as we already do) and if the state is differently, we assume, the message should just be sent as in the past.

and if the given chat_id differs from the chat_id the message was preprared for, maybe just call this an error.

Reply:
According to the docs of `dc_msg_t`:

> The message object is not updated.

Although it is not strictly true, since the `id`, `chat_id` and `param` are updated by `dc_send_msg()`, `state` is indeed not changed (neither before nor after this PR). The check for `msg->state` in the code is actually a bug and passes the tests only because the Python bindings always return a new object loaded from the DB after the C call returns.

Theoretically, we should either change the "not updated" part of the doc and clarify *what* and *when* we change, or stop modifying `dc_msg_t` parameters. Depending on that decision (not necessarily waiting until the decision is fully implemented) we can either use `msg->state` instead of `chat_id`, or remove the check for `state`. And/or use something completely different to tell whether `dc_prepare_msg()` was already called (worst case: have `dc_send_msg()` look up the state in the DB).

Reply:
> The message object is not updated.

this is more a warning that if you have a dc_msg_t object and the state changes in the database, eg. by receiving a read-receipt, the message objects are not updated.
maybe we should reword this part, however.

but i think it is fine to set the id and the state of a dc_msg_t object on sending - as mentioned, it already happens for some other field and the object is not declared as `const`.

@VP- apart from that: is this pr ready to merge from your point of view? i'd like to get it to master this week.